src/main/java/org/dataloader/impl/DataLoaderImpl.java

```diff
@@ -85,33 +222,32 @@ Object getCallContext() {
         return Optional.empty();
     }
 
-    Optional<CompletableFuture<V>> getIfCompleted(K key) {
-        synchronized (dataLoader) {
-            Optional<CompletableFuture<V>> cachedPromise = getIfPresent(key);
-            if (cachedPromise.isPresent()) {
-                CompletableFuture<V> promise = cachedPromise.get();
-                if (promise.isDone()) {
-                    return cachedPromise;
-                }
-            }
-        }
-        return Optional.empty();
+    Optional<CompletableFuture<V>> getIfCompletedImpl(K key) {
+        return getIfPresent(key).filter(CompletableFuture::isDone);
     }
 
+    CompletableFuture<V> loadImpl(K key, Object loadContext) {
+        Object cacheKey = getCacheKey(nonNull(key));
+        boolean batchingEnabled = loaderOptions.batchingEnabled();
+        boolean cachingEnabled = loaderOptions.cachingEnabled();
 
-    CompletableFuture<V> load(K key, Object loadContext) {
-        synchronized (dataLoader) {
-            Object cacheKey = getCacheKey(nonNull(key));
+        synchronized (this) {
```

Comment:
Say we have caching and batching disabled and the batchFunction happens to have some blocking code (not ideal). Having this entire block `synchronized` would cause to only have at most one `batchFunction` call executed instead per data loader instance, no?

Do you see any risk of contention?